core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pmic.c

```diff
@@ -0,0 +1,556 @@
+
+/* Expect a single PMIC instance */
+static struct i2c_handle_s i2c_handle;
+static uint32_t pmic_i2c_addr;
+
+bool stm32mp_with_pmic(void)
+{
+	return i2c_handle.dt_status & DT_STATUS_OK_SEC;
+}
+
+static int dt_get_pmic_node(void *fdt)
+{
+	return fdt_node_offset_by_compatible(fdt, -1, "st,stpmic1");
+}
+
+static int dt_pmic_status(void)
+{
+	void *fdt = get_embedded_dt();
+
+	if (fdt) {
+		int node = dt_get_pmic_node(fdt);
+
+		if (node > 0)
+			return _fdt_get_status(fdt, node);
+	}
+
+	return -1;
+}
+
+static bool dt_pmic_is_secure(void)
+{
+	int status = dt_pmic_status();
+
+	return status == DT_STATUS_OK_SEC &&
+	       i2c_handle.dt_status == DT_STATUS_OK_SEC;
+}
+
+/*
+ * @idx: Private identifier provided by the target PMIC driver
+ * @flags: Operations expected when entering a low power sequence
+ * @voltage: Target voltage to apply during low power sequences
+ */
+struct regu_bo_config {
+	uint8_t flags;
+	struct stpmic1_bo_cfg cfg;
+};
+
+#define REGU_BO_FLAG_ENABLE_REGU		BIT(0)
+#define REGU_BO_FLAG_SET_VOLTAGE		BIT(1)
+#define REGU_BO_FLAG_PULL_DOWN			BIT(2)
+#define REGU_BO_FLAG_MASK_RESET			BIT(3)
+
+static struct regu_bo_config *regu_bo_config;
+static size_t regu_bo_count;
+
+static int save_boot_on_config(void)
+{
+	int pmic_node = 0;
+	int regulators_node = 0;
+	int regu_node = 0;
+	void *fdt = NULL;
+
+	assert(!regu_bo_config && !regu_bo_count);
+
+	fdt = get_embedded_dt();
+	if (!fdt)
+		panic();
+
+	pmic_node = dt_get_pmic_node(fdt);
+	if (pmic_node < 0)
+		panic();
+
+	regulators_node = fdt_subnode_offset(fdt, pmic_node, "regulators");
+
+	fdt_for_each_subnode(regu_node, fdt, regulators_node) {
+		const fdt32_t *cuint = NULL;
+		const char *name = NULL;
+		struct regu_bo_config regu_cfg = { };
+		uint16_t mv = 0;
+
+		if (_fdt_get_status(fdt, regu_node) == DT_STATUS_DISABLED)
+			continue;
+
+		if (!fdt_getprop(fdt, regu_node, "regulator-boot-on", NULL))
+			continue;
+
+		memset(&regu_cfg, 0, sizeof(regu_cfg));
+		name = fdt_get_name(fdt, regu_node, NULL);
+
+		regu_cfg.flags |= REGU_BO_FLAG_ENABLE_REGU;
+
+		if (fdt_getprop(fdt, regu_node, "regulator-pull-down", NULL)) {
+			stpmic1_bo_pull_down_cfg(name, &regu_cfg.cfg);
+			regu_cfg.flags |= REGU_BO_FLAG_PULL_DOWN;
+		}
+
+		if (fdt_getprop(fdt, regu_node, "st,mask-reset", NULL)) {
+			stpmic1_bo_mask_reset_cfg(name, &regu_cfg.cfg);
+			regu_cfg.flags |= REGU_BO_FLAG_MASK_RESET;
+		}
+
+		cuint = fdt_getprop(fdt, regu_node,
+				    "regulator-min-microvolt", NULL);
+		if (cuint) {
+			/* DT uses microvolts and driver awaits millivolts */
+			mv = (uint16_t)(fdt32_to_cpu(*cuint) / 1000U);
+
+			if (!stpmic1_bo_voltage_cfg(name, mv, &regu_cfg.cfg))
+				regu_cfg.flags |= REGU_BO_FLAG_SET_VOLTAGE;
+		}
+
+		/* Save config in the Boot On configuration list */
+		regu_bo_count++;
+		regu_bo_config = realloc(regu_bo_config,
+					 regu_bo_count * sizeof(regu_cfg));
+		if (!regu_bo_config)
+			panic();
+
+		memcpy(&regu_bo_config[regu_bo_count - 1], &regu_cfg,
```

Comment:
This could be written more conveniently as `regu_bo_config[regu_bo_count - 1] = regu_cfg`, as a bonus you get type safety.

Reply:
ok, will do.